src/Unleash/Internal/CachedFilesLoader.cs

```diff
@@ -13,102 +14,138 @@ internal class CachedFilesLoader
-                    eventConfig?.RaiseError(new ErrorEvent() { Error = ex, ErrorType = ErrorType.FileCache });
-                }
+                return SafeRead(legacyPath, defaultContent);
             }
 
-            if (result.InitialToggleCollection == null)
+            return SafeCreateAndReturn(primaryPath, defaultContent);
+        }
+
+        private T EnsureFileAndReadJsonWithLegacy<T>(string primaryPath, string legacyPath) where T : class
+        {
+            if (fileSystem.FileExists(primaryPath))
             {
-                result.InitialETag = string.Empty;
+                return SafeReadJson<T>(primaryPath);
             }
 
-            if ((result.InitialToggleCollection == null || result.InitialToggleCollection.Features?.Count == 0 || bootstrapOverride) && toggleBootstrapProvider != null)
+            if (fileSystem.FileExists(legacyPath))
             {
-                var bootstrapCollection = toggleBootstrapProvider.Read();
-                if (bootstrapCollection != null && bootstrapCollection.Features?.Count > 0)
-                    result.InitialToggleCollection = bootstrapCollection;
+                return SafeReadJson<T>(legacyPath);
             }
 
-            return result;
+            try
+            {
+                fileSystem.WriteAllText(primaryPath, string.Empty);
+            }
+            catch (IOException ex)
+            {
+                // Should get handled later when we try to write
+                Logger.Debug(() => $"UNLEASH: Failed to create backup file: {primaryPath}", ex);
+            }
+            return null;
+        }
+
+        private string SafeRead(string path, string defaultContent)
+        {
+            try { return fileSystem.ReadAllText(path); }
+            catch (IOException ex)
+            {
+                Logger.Warn(() => $"UNLEASH: Failed to read etag file '{path}'.", ex);
+                return defaultContent;
+            }
+        }
+
+        private string SafeCreateAndReturn(string path, string defaultContent)
+        {
+            try
+            {
+                fileSystem.WriteAllText(path, defaultContent);
+            }
+            catch (IOException ex)
+            {
+                Logger.Warn(() => $"UNLEASH: Failed to create backup file: {path}", ex);
+            }
+            return defaultContent;
+        }
+
+        private T SafeReadJson<T>(string path) where T : class
+        {
+            try
+            {
+                using (var stream = fileSystem.FileOpenRead(path))
+                {
+                    return jsonSerializer.Deserialize<T>(stream);
+                }
+            }
+            catch (IOException ex)
+            {
+                Logger.Warn(() => $"UNLEASH: Failed to load backup file: {path}", ex);
+                eventConfig?.RaiseError(new ErrorEvent() { ErrorType = ErrorType.FileCache, Error = ex });
+                return null;
+            }
+        }
```

Comment:
All of these seem almost generic enough to belong somewhere else, like `FileSystem.cs`. Is that worth considering?